File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> result;
        vector<string> chr_l;
        
        if (digits.empty()) {
            return result;
        }

        // 处理单个数字的情况
        if (digits.size() == 1) {
            int num = digits[0] - '0'; // 将字符转换为数字
            if (num < 7) {
                for (int i = 0; i < 3; i++) {
                    result.push_back(string(1, char((97 + (num - 2) * 3) + i)));
                }
            } else if (num == 7) {
                for (int i = 0; i < 4; i++) {
                    result.push_back(string(1, char(112 + i)));
                }
            } else if (num == 8) {
                for (int i = 0; i < 3; i++) {
                    result.push_back(string(1, char(116 + i)));
                }
            } else if (num == 9) {
                for (int i = 0; i < 4; i++) {
                    result.push_back(string(1, char(119 + i)));
                }
            }
            return result;
        }

        // 对多个数字进行处理
        for (char d : digits) {
            string all_chr = "";
            int num = d - '0'; // 将字符转换为数字
            if (num < 7) {
                for (int i = 0; i < 3; i++) {
                    all_chr += char((97 + (num - 2) * 3) + i);
                }
            } else if (num == 7) {
                for (int i = 0; i < 4; i++) {
                    all_chr += char(112 + i);
                }
            } else if (num == 8) {
                for (int i = 0; i < 3; i++) {
                    all_chr += char(116 + i);
                }
            } else if (num == 9) {
                for (int i = 0; i < 4; i++) {
                    all_chr += char(119 + i);
                }
            }
            chr_l.push_back(all_chr);
        }

        // 根据输入的数字组合出所有可能的字母组合
        if (chr_l.size() == 2) {
            for (char c : chr_l[0]) {
                for (char sub_ : chr_l[1]) {
                    result.push_back(string(1, c) + sub_);
                }
            }
        } else if (chr_l.size() == 3) {
            for (char c : chr_l[0]) {
                for (char sub_ : chr_l[1]) {
                    for (char ss : chr_l[2]) {
                        result.push_back(string(1, c) + sub_ + ss);
                    }
                }
            }
        } else if (chr_l.size() == 4) {
            for (char c : chr_l[0]) {
                for (char sub_ : chr_l[1]) {
                    for (char ss : chr_l[2]) {
                        for (char sss : chr_l[3]) {
                            result.push_back(string(1, c) + sub_ + ss + sss);
                        }
                    }
                }
            }
        }

        return result;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        result = []
        chr_l = []
        if len(digits)==0:
            return []
        if len(digits)==1:
            num = int(digits)
            if num<7:
                for i in range(0,3):
                    result.append(chr((97+(num-2)*3)+i))
                return result
            elif num==7:
                for i in range(0,4):
                    result.append(chr(112+i))
                return result
            elif num==8:
                for i in range(0,3):
                    result.append(chr(116+i))
                return result
            elif num==9:
                for i in range(0,4):
                    result.append(chr(119+i))
                return result
        else:
            for d in digits:
                all_chr = ""
                num = int(d)
                if num<7:
                    for i in range(0,3):
                        all_chr+=(chr((97+(num-2)*3)+i))
                elif num==7:
                    for i in range(0,4):
                        all_chr+=(chr(112+i))
                elif num==8:
                    for i in range(0,3):
                        all_chr+=(chr(116+i))
                elif num==9:
                    for i in range(0,4):
                        all_chr+=(chr(119+i))
                chr_l.append(all_chr)
            
            if len(chr_l)==2:
                for c in chr_l[0]:
                    for sub_ in chr_l[1]:
                        result.append(str(c)+str(sub_))
            elif len(chr_l)==3:
                for c in chr_l[0]:
                    for sub_ in chr_l[1]:
                        for ss in chr_l[2]:
                            result.append(str(c)+str(sub_)+str(ss))
            else:
                for c in chr_l[0]:
                    for sub_ in chr_l[1]:
                        for ss in chr_l[2]:
                            for sss in chr_l[3]:
                                result.append(str(c)+str(sub_)+ str(ss)+ str(sss))
            return result

        
